### ISSUE TITLE
Image ssh-agent:4.1.0-jdk8 needs privileged mode on docker 18.03

### ISSUE THREAD

**Issue by grostarin** [jenkinsci/docker-ssh-agent#86]
### Version report

Jenkins and plugins versions report:

Jenkins: 2.289.3
OS: Linux - 3.10.0-1127.8.2.el7.x86_64
---
ace-editor:1.1
analysis-model-api:10.3.0
ant:1.11
antisamy-markup-formatter:2.1
apache-httpcomponents-client-4-api:4.5.13-1.0
authentication-tokens:1.4
batch-task:1.19
bootstrap4-api:4.6.0-3
bootstrap5-api:5.1.1-1
bouncycastle-api:2.24
branch-api:2.7.0
build-monitor-plugin:1.12+build.201809061734
buildtriggerbadge:2.11
caffeine-api:2.9.2-29.v717aac953ff3
checks-api:1.7.2
chucknorris:1.4
cloudbees-folder:6.16
cloverphp:0.6
command-launcher:1.6
conditional-buildstep:1.4.1
config-autorefresh-plugin:1.0
config-file-provider:3.8.1
configurationslicing:1.52
credentials:2.6.1
credentials-binding:1.27
cvs:2.19
data-tables-api:1.11.2-1
delivery-pipeline-plugin:1.4.2
dependency-check-jenkins-plugin:5.1.1
dependency-track:4.0.0
display-url-api:2.3.5
docker-commons:1.17
docker-java-api:3.1.5.2
docker-plugin:1.2.3
docker-swarm:1.11
docker-workflow:1.26
dtkit-api:3.0.0
durable-task:1.39
echarts-api:5.1.2-11
email-ext:2.83
extended-choice-parameter:0.82
extended-read-permission:3.2
external-monitor-job:1.7
font-awesome-api:5.15.4-1
forensics-api:1.3.1
gatling:1.3.0
git:4.8.2
git-client:3.9.0
git-parameter:0.9.13
git-server:1.10
gitlab-plugin:1.5.20
global-build-stats:1.5
global-variable-string-parameter:1.2
gradle:1.37.1
gravatar:2.2
greenballs:1.15.1
groovy:2.4
h2-api:1.4.199
handlebars:3.0.8
hidden-parameter:0.0.4
htmlpublisher:1.25
jackson2-api:2.12.4
javadoc:1.6
jdepend:1.3.0
jdk-tool:1.5
jobConfigHistory:2.28.1
jquery:1.12.4-1
jquery-detached:1.2.1
jquery3-api:3.6.0-2
jsch:0.1.55.2
junit:1.53
ldap:2.7
lockable-resources:2.11
m2release:0.16.2
mailer:1.34
managed-scripts:1.5.4
mapdb-api:1.0.9.0
matrix-auth:2.6.8
matrix-project:1.19
maven-metadata-plugin:2.0.0
maven-plugin:3.12
momentjs:1.1.1
monitoring:1.88.0
nexus-artifact-uploader:2.13
pam-auth:1.6
parameterized-trigger:2.41
phing:0.13.3
pipeline-build-step:2.15
pipeline-graph-analysis:1.11
pipeline-input-step:2.12
pipeline-maven:3.10.0
pipeline-milestone-step:1.3.2
pipeline-model-api:1.9.2
pipeline-model-definition:1.9.2
pipeline-model-extensions:1.9.2
pipeline-rest-api:2.19
pipeline-stage-step:2.5
pipeline-stage-tags-metadata:1.9.2
pipeline-stage-view:2.19
pipeline-utility-steps:2.10.0
plain-credentials:1.7
plot:2.1.9
plugin-util-api:2.4.0
popper-api:1.16.1-2
popper2-api:2.10.1-1
project-inheritance:21.04.03
promoted-builds:3.10
publish-over:0.22
publish-over-ssh:1.22
rebuild:1.32
repository-connector:2.0.5
role-strategy:3.2.0
run-condition:1.5
scm-api:2.6.5
script-security:1.78
snakeyaml-api:1.29.1
sonar:2.13.1
ssh-credentials:1.19
ssh-slaves:1.33.0
sshd:3.1.0
structs:1.23
subversion:2.14.4
token-macro:266.v44a80cf277fd
translation:1.16
trilead-api:1.0.13
uno-choice:2.5.6
warnings-ng:9.5.0
windows-slaves:1.8
workflow-aggregator:2.6
workflow-api:2.46
workflow-basic-steps:2.24
workflow-cps:2.94
workflow-cps-global-lib:2.21
workflow-durable-task-step:2.40
workflow-job:2.41
workflow-multibranch:2.26
workflow-scm-step:2.13
workflow-step-api:2.24
workflow-support:3.8
xunit:3.0.3

- What Operating System are you using (both controller, and any agents involved in the problem)?

Jenkins controller and agent running on same linux VM : RHEL7
Everything runs using docker version 18.03.1-ce, build 9ee9f40

### Reproduction steps

run ssh agent (same behavior on VM and on my docker desktop up to date on my laptop)

> docker run -d -e "JENKINS_AGENT_SSH_PUBKEY=<SSH_PUB_KEY>" jenkins/ssh-agent:4.1.0-jdk8

display logs on created container (do it quick because it will disappear in few seconds)

> docker logs <CONTAINER_ID> -f

### Results

Expected result:

SSH is running.

In container log...
> ... (init script)
> Server listening on 0.0.0.0 port 22.
> Server listening on :: port 22.

Actual result:

Failure => SSH is not running

In container log...

> + [[ <SSH_PUB_KEY> == ssh-* ]]
> + write_key '<SSH_PUB_KEY>'
> + local ID_GROUP
> ++ stat -c %U:%G /home/jenkins
> stat: cannot statx '/home/jenkins': Operation not permitted
> + ID_GROUP=

My investigation :

I think issue is linked to call using linux command stat here : https://github.com/jenkinsci/docker-ssh-agent/blob/bf132a036688e08699dab9427fdd749183695053/setup-sshd#L38
It seems that only a privileged container could successfully run this command.

Using --privileged in docker run solve the problem :
> docker run -d --privileged -e "JENKINS_AGENT_SSH_PUBKEY=<SSH_PUB_KEY>" jenkins/ssh-agent:4.1.0-jdk8

**Comment by MarkEWaite** [comment]
I was unable to duplicate the problem you're describing with the commands that you provided.  The ssh agent process starts and listens on its internal port 22 when I use the command:

```
$ docker run  -d -e "JENKINS_AGENT_SSH_PUBKEY=$PUBLIC_KEY" jenkins/ssh-agent:4.1.0-jdk8
```

I confirmed that the `stat` command is allowed for an unprivileged user on my Ubuntu Linux computer and my OpenBSD computer.

Maybe you're using some form of mount that causes the `/home/jenkins` directory to have the wrong permissions or to require some additional permissions?

**Comment by grostarin** [comment]
I just run this command :

> docker run  -d -e "JENKINS_AGENT_SSH_PUBKEY=$PUBLIC_KEY" jenkins/ssh-agent:4.1.0-jdk8

No volume mounted (except those automatically created by docker because they is some volume declared in this image build).
I made so few tests, because it makes me crazy ;)

I fail on two VM RHEL (7.8, 7.9) - docker-ce 18.03.1.
It is ok on : 
- windows 10 docker-ce 20.10.8
- windows 10 wsl ubuntu 20.04 docker-ce 20.10.8

It's probably a bug in docker-ce which has been fixed.
I'll override setup-sshd to avoir stat call and use jenkins:jenkins directly or from env variables...

I'm sad ;)
